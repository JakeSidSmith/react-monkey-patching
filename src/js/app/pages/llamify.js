import React from 'react';
import applyPatch from '../../monkey-patches/apply-patch';

class StringClass extends React.PureComponent {
  render () {
    return 'String class';
  }
}

class ArrayClass extends React.PureComponent {
  render () {
    return [
      [
        'Array',
        'class'
      ]
    ];
  }
}

const StringSFC = () => (
  'String SFC'
);

const ArraySFC = () => (
  [
    'Array',
    'SFC'
  ]
);

export default class Llamify extends React.PureComponent {
  constructor (props) {
    super(props);

    this.state = {
      value: ''
    };

    this.onChange = this.onChange.bind(this);
  }

  onChange (event) {
    this.setState({
      value: event.target.value
    });
  }

  onApplyPatch () {
    applyPatch('llamify');
  }

  render () {
    const { value } = this.state;

    return (
      <div>
        <button data-no-patch onClick={this.onApplyPatch}>
          Apply monkey patch
        </button>
        <p>
          Hello, World!
        </p>
        <p>
          <StringClass />
        </p>
        <p>
          <ArrayClass />
        </p>
        <p>
          <StringSFC />
        </p>
        <p>
          <ArraySFC />
        </p>
        <p>
          <input type="submit" value="Input type = submit" />
        </p>
        <p>
          <input type="button" value="Input type = button" />
        </p>
        <p>
          <input
            type="text"
            placeholder="Input type = text placeholder"
            value={value}
            onChange={this.onChange}
          />
        </p>
        <p>
          <img alt="Alt text" />
        </p>
        <p>
          <a href="#" title="Link title">
            Link with title
          </a>
        </p>
        <p>
          Alt number
          <img alt={1} />
        </p>
        <p>
          <a href="#" title={1}>
            Link with title that's a number
          </a>
        </p>
        <p>
          Short word that doesn't need scrambling: "hi"
        </p>
      </div>
    );
  }
}
