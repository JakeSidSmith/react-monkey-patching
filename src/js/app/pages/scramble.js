import React from 'react';

export default class Scramble extends React.PureComponent {
  render () {
    return (
      <div>
        Scramble
      </div>
    );
  }
}
