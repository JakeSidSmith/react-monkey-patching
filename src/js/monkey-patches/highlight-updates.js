import React from 'react';
import ReactDOM from 'react-dom';

const patchedClasses = [];
const patchedMethods = [];

const HIGHLIGHT_CLASS = 'react-update-highlight';
const MATCHES_HIGHLIGHT_CLASS = /\s*\breact-update-highlight\b\s*/gi;
const DEFAULT_COLOR = 'rgba(255, 0, 0, 0.5)';

const STYLE = document.createElement('style');
STYLE.type = 'text/css';

const createStyles = (color = DEFAULT_COLOR) => {
  return `.react-update-highlight {
    animation-name: react-update-highlight;
    animation-duration: 2s;
    animation-iteration-count: 1;
  }
  @keyframes react-update-highlight {
    from {
      background-color: ${color};
    }
    to {
      background-color: transparent;
    }
  }`;
};

STYLE.innerHTML = createStyles();

document.head.appendChild(STYLE);

const highlightUpdates = (color = DEFAULT_COLOR) => {
  STYLE.innerHTML = createStyles(color);

  const originalCreateElement = React.createElement;

  React.createElement = function () {
    const args = Array.prototype.slice.call(arguments);
    const component = args[0];

    if (
      typeof component === 'function' &&
      component.prototype instanceof React.Component &&
      patchedClasses.indexOf(component) < 0
    ) {
      const originalComponentDidUpdate = component.prototype.componentDidUpdate;

      patchedClasses.push(component);
      patchedMethods.push(originalComponentDidUpdate);

      component.prototype.hasReactUpdateHighlight = true;
      component.prototype.componentDidUpdate = function () {
        const updateArgs = Array.prototype.slice.call(arguments);
        let node;

        try {
          node = ReactDOM.findDOMNode(this);
        } catch (error) {
          return;
        }

        if (!node) {
          return;
        }

        if (MATCHES_HIGHLIGHT_CLASS.test(node.className)) {
          node.className = node.className.replace(MATCHES_HIGHLIGHT_CLASS, '');
        }

        void node.offsetWidth; // tslint:disable-line:no-unused-expression
        const needsSpace = node.className || node.className.lastIndexOf(' ') !== node.className.length - 1;
        node.className = `${node.className}${needsSpace ? ' ' + HIGHLIGHT_CLASS : HIGHLIGHT_CLASS}`;

        if (typeof originalComponentDidUpdate === 'function') {
          originalComponentDidUpdate.apply(this, updateArgs);
        }
      };
    }

    return originalCreateElement.apply(this, args);
  };

  return () => {
    React.createElement = originalCreateElement;

    while (patchedClasses.length) {
      const patchedClass = patchedClasses.shift();
      patchedClass.prototype.componentDidUpdate = patchedMethods.shift();
    }
  };
};

export default () => highlightUpdates();
