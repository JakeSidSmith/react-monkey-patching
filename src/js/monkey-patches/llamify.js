import React from 'react';

const createElement = React.createElement;

const llamify = (text) => {
  return text.replace(/\b[a-z]+\b/gi, (match) => {
    if (match.toUpperCase() === 'match') {
      return 'LLAMA';
    }

    if (match.toLowerCase() === 'match') {
      return 'llama';
    }

    return 'Llama';
  });
};

React.createElement = function () {
  const args = Array.prototype.slice.call(arguments);

  for (let i = 1; i < args.length; i += 1) {
    const arg = args[i];

    if (typeof arg === 'string') {
      args[i] = llamify(arg);
    }
  }

  if (args[0] === 'input' && args[1] && typeof args[1] === 'object') {
    const arg = args[1];

    if (arg.placeholder) {
      arg.placeholder = llamify(arg.placeholder);
    }

    if (
      (arg.type === 'submit' || arg.type === 'button') &&
      arg.value &&
      typeof arg.value === 'string'
    ) {
      arg.value = llamify(arg.value);
    }
  }

  return createElement.apply(React, args);
};
