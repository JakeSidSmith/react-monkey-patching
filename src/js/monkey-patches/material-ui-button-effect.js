import React from 'react';

const STYLE = document.createElement('style');
STYLE.type = 'text/css';

STYLE.innerHTML = `
  .material-ui-button-effect {
    position: absolute;
    top: 50%;
    left: 50%;
    margin-top: -50%;
    margin-left: -50%;
    width: 100%;
    height: 0;
    padding-bottom: 100%;
    border-radius: 50%;
    background-color: #FFFFFF;
    mix-blend-mode: overlay;
    animation-name: material-ui-button-effect;
    animation-duration: 1s;
    animation-iteration-count: 1;
    animation-timing-function: ease-out;
    animation-fill-mode: forwards;
  }
  @keyframes material-ui-button-effect {
    from {
      opacity: 0.5;
      transform: scale(0, 0);
    }
    to {
      opacity: 0;
      transform: scale(2, 2);
    }
  }
`;

document.head.appendChild(STYLE);

const materialUIButtonEffect = () => {
  const originalCreateElement = React.createElement;

  React.createElement = function createElement () {
    const args = Array.prototype.slice.call(arguments);

    if (!args[1]) {
      args[1] = {};
    }

    const element = args[0];
    const props = args[1];

    if (props && props['data-no-patch']) {
      return originalCreateElement.apply(React, args);
    }

    if (element === 'button') {
      const originalOnClick = props.onClick;

      props.style = Object.assign({}, {position: 'relative', overflow: 'hidden'}, props.style);

      props.onClick = function (event) {
        const onClickArgs = Array.prototype.slice.call(arguments);

        const node = event.currentTarget;

        if (!node) {
          return;
        }

        const pew = document.createElement('div');
        pew.className = 'material-ui-button-effect';

        node.appendChild(pew);

        setTimeout(() => {
          if (pew) {
            pew.remove();
          }
        }, 2000);

        if (typeof originalOnClick === 'function') {
          return originalOnClick.apply(this, onClickArgs);
        }
      };
    }

    return originalCreateElement.apply(React, args);
  };

  return () => {
    React.createElement = originalCreateElement;
  };
};

export default materialUIButtonEffect;
