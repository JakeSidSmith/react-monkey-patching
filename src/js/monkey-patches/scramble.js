import { patch } from './patch';

const wordFinder = /\b[a-z]+\b/gi;

const scrambleWord = (word) => {
  if (word.length <= 2) {
    return word;
  }

  const newWord = word.split('');
  const firstLetter = newWord.shift();
  const lastLetter = newWord.pop();

  for (let i = 0; i < word.length; i += 1) {
    const from = Math.floor(Math.random() * word.length);
    const to = Math.floor(Math.random() * word.length);
    newWord.splice(to, 0, newWord.splice(from, 1)[0]);
  }

  newWord.unshift(firstLetter);
  newWord.push(lastLetter);

  return newWord.join('');
};

const scrambleWords = (text) => {
  return text.replace(wordFinder, scrambleWord);
};

patch(scrambleWords);
