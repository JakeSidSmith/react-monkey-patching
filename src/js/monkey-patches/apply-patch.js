import llamify from './llamify';

const PATCHES = {
  llamify
};

let unpatch;

const applyPatch = (name) => {
  if (typeof unpatch === 'function') {
    unpatch();
    unpatch = null;
  }

  unpatch = PATCHES[name]();
};

export default applyPatch;
