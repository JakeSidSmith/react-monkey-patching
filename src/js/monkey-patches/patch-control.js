import llamify from './llamify';
import scramble from './scramble';
import highlightUpdates from './highlight-updates';
import materialUIButtonEffect from './material-ui-button-effect';


const PATCHES = {
  llamify,
  scramble,
  highlightUpdates,
  materialUIButtonEffect
};

const subscriptions = [];

export const subscribe = (callback) => {
  if (typeof callback === 'function' && subscriptions.indexOf(callback) < 0) {
    subscriptions.push(callback);
  }

  return () => {
    const index = subscriptions.indexOf(callback);

    if (index >= 0) {
      subscriptions.splice(index, 1);
    }
  };
};

const updateComponents = () => {
  subscriptions.forEach((callback) => {
    callback();
  });
};

let unpatch;

export const removePatch = () => {
  if (typeof unpatch === 'function') {
    unpatch();
  }

  updateComponents();
};

export const applyPatch = (name) => {
  if (typeof unpatch === 'function') {
    unpatch();
    unpatch = null;
  }

  unpatch = PATCHES[name]();
  updateComponents();
};
