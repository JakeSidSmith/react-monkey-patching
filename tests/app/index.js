import React from 'react';
import renderer from 'react-test-renderer';

import App from '../../src/js/app/';

describe('App', () => {

  const originalCreateElement = React.createElement;

  beforeEach(() => {
    React.createElement = originalCreateElement;
  });

  it('should match snapshot', () => {
    expect(renderer.create(<App />)).toMatchSnapshot();
  });

  it('should match snapshot when llamified', () => {
    require('../../src/js/monkey-patches/llamify');

    expect(renderer.create(<App />)).toMatchSnapshot();
  });

  it('should match snapshot when scrambled', () => {
    require('../../src/js/monkey-patches/scramble');

    expect(renderer.create(<App />)).toMatchSnapshot();
  });

});
