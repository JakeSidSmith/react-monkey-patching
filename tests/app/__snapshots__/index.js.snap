// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<table
  className="app"
>
  <tbody>
    <tr>
      <td
        className="navigation"
      >
        <div
          className="scrollable"
        >
          <ul>
            <li>
              <a
                href="#/"
                onClick={[Function]}
              >
                Intro
              </a>
            </li>
            <li>
              <a
                href="#/what-is-monkey-patching/"
                onClick={[Function]}
              >
                What is monkey patching?
              </a>
            </li>
            <li>
              <a
                href="#/why-react-create-element/"
                onClick={[Function]}
              >
                Why React?
              </a>
            </li>
            <li>
              <a
                href="#/react-highlight-updates/"
                onClick={[Function]}
              >
                React highlight updates
              </a>
            </li>
            <li>
              <a
                href="#/material-ui-button-effect/"
                onClick={[Function]}
              >
                Material UI button effect
              </a>
            </li>
            <li>
              <a
                href="#/react-fastclick/"
                onClick={[Function]}
              >
                React fastclick
              </a>
            </li>
            <li>
              <a
                href="#/simple-patch-function/"
                onClick={[Function]}
              >
                Simple patch function
              </a>
            </li>
            <li>
              <a
                href="#/llamify/"
                onClick={[Function]}
              >
                Llamify
              </a>
            </li>
            <li>
              <a
                href="#/scramble/"
                onClick={[Function]}
              >
                Scramble
              </a>
            </li>
            <li>
              <a
                href="#/internationalization/"
                onClick={[Function]}
              >
                Internationalization
              </a>
            </li>
          </ul>
        </div>
      </td>
      <td
        className="main"
      >
        <div>
          <h1
            className="page-title"
          >
            Intro
          </h1>
          <div
            className="scrollable"
          >
            <div
              className="container"
            >
              <div
                className="vertical-center"
              >
                <div>
                  <h2>
                    Monkey Patching React
                  </h2>
                  <p>
                    Hello! My name is Jake, and today I'm going to show you some totally not hacks at all, believe me, I promise, really not hacks... awesome monkey patching of React.
                  </p>
                  <h2>
                    Want more awesome stuff?
                  </h2>
                  <p>
                    Here's some links you might enjoy:
                  </p>
                  <ul>
                    <li>
                      <a
                        href="https://jakesidsmith.github.io/react-monkey-patching/"
                        target="_blank"
                        title="This demo"
                      >
                        This demo - jakesidsmith.github.io/react-monkey-patching/
                      </a>
                    </li>
                    <li>
                      <a
                        href="http://jakesidsmith.com"
                        target="_blank"
                        title="My Blog"
                      >
                        My Blog - jakesidsmith.com
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://github.com/jakesidsmith/"
                        target="_blank"
                        title="My GitHub"
                      >
                        My GitHub - github.com/jakesidsmith/
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://twitter.com/jakesidsmith/"
                        target="_blank"
                        title="My Twitter"
                      >
                        My Twitter - twitter.com/jakesidsmith/
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://www.npmjs.com/package/react-highlight-updates"
                        target="_blank"
                        title="React Highlight Updates"
                      >
                        React Highlight Updates (we'll cover this shortly) - npmjs.com/package/react-highlight-updates
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://www.npmjs.com/package/react-fastclick"
                        target="_blank"
                        title="React Fastclick"
                      >
                        React Fastclick (we'll cover this briefly also) - npmjs.com/package/react-fastclick
                      </a>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`App should match snapshot when llamified 1`] = `
<table
  className="app"
>
  <tbody>
    <tr>
      <td
        className="navigation"
      >
        <div
          className="scrollable"
        >
          <ul>
            <li>
              <a
                href="#/"
                onClick={[Function]}
              >
                Llama
              </a>
            </li>
            <li>
              <a
                href="#/what-is-monkey-patching/"
                onClick={[Function]}
              >
                Llama llama llama llama?
              </a>
            </li>
            <li>
              <a
                href="#/why-react-create-element/"
                onClick={[Function]}
              >
                Llama Llama?
              </a>
            </li>
            <li>
              <a
                href="#/react-highlight-updates/"
                onClick={[Function]}
              >
                Llama llama llama
              </a>
            </li>
            <li>
              <a
                href="#/material-ui-button-effect/"
                onClick={[Function]}
              >
                Llama LLAMA llama llama
              </a>
            </li>
            <li>
              <a
                href="#/react-fastclick/"
                onClick={[Function]}
              >
                Llama llama
              </a>
            </li>
            <li>
              <a
                href="#/simple-patch-function/"
                onClick={[Function]}
              >
                Llama llama llama
              </a>
            </li>
            <li>
              <a
                href="#/llamify/"
                onClick={[Function]}
              >
                Llama
              </a>
            </li>
            <li>
              <a
                href="#/scramble/"
                onClick={[Function]}
              >
                Llama
              </a>
            </li>
            <li>
              <a
                href="#/internationalization/"
                onClick={[Function]}
              >
                Llama
              </a>
            </li>
          </ul>
        </div>
      </td>
      <td
        className="main"
      >
        <div>
          <h1
            className="page-title"
          >
            Llama
          </h1>
          <div
            className="scrollable"
          >
            <div
              className="container"
            >
              <div
                className="vertical-center"
              >
                <div>
                  <h2>
                    Llama Llama Llama
                  </h2>
                  <p>
                    Llama! Llama llama llama Llama, llama llama LLAMA'llama llama llama llama llama llama llama llama llama llama llama, llama llama, LLAMA llama, llama llama llama... llama llama llama llama Llama.
                  </p>
                  <h2>
                    Llama llama llama llama?
                  </h2>
                  <p>
                    Llama'llama llama llama llama llama llama:
                  </p>
                  <ul>
                    <li>
                      <a
                        href="https://jakesidsmith.github.io/react-monkey-patching/"
                        target="_blank"
                        title="Llama llama"
                      >
                        Llama llama - llama.llama.llama/llama-llama-llama/
                      </a>
                    </li>
                    <li>
                      <a
                        href="http://jakesidsmith.com"
                        target="_blank"
                        title="Llama Llama"
                      >
                        Llama Llama - llama.llama
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://github.com/jakesidsmith/"
                        target="_blank"
                        title="Llama Llama"
                      >
                        Llama Llama - llama.llama/llama/
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://twitter.com/jakesidsmith/"
                        target="_blank"
                        title="Llama Llama"
                      >
                        Llama Llama - llama.llama/llama/
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://www.npmjs.com/package/react-highlight-updates"
                        target="_blank"
                        title="Llama Llama Llama"
                      >
                        Llama Llama Llama (llama'llama llama llama llama) - llama.llama/llama/llama-llama-llama
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://www.npmjs.com/package/react-fastclick"
                        target="_blank"
                        title="Llama Llama"
                      >
                        Llama Llama (llama'llama llama llama llama llama) - llama.llama/llama/llama-llama
                      </a>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`App should match snapshot when scrambled 1`] = `
<table
  className="app"
>
  <tbody>
    <tr>
      <td
        className="navigation"
      >
        <div
          className="scrollable"
        >
          <ul>
            <li>
              <a
                href="#/"
                onClick={[Function]}
              >
                Inrto
              </a>
            </li>
            <li>
              <a
                href="#/what-is-monkey-patching/"
                onClick={[Function]}
              >
                Waht is moenky phcitang?
              </a>
            </li>
            <li>
              <a
                href="#/why-react-create-element/"
                onClick={[Function]}
              >
                Why Recat?
              </a>
            </li>
            <li>
              <a
                href="#/react-highlight-updates/"
                onClick={[Function]}
              >
                Rcaet hghigilht uadepts
              </a>
            </li>
            <li>
              <a
                href="#/material-ui-button-effect/"
                onClick={[Function]}
              >
                Meriaatl UI bttoun efcfet
              </a>
            </li>
            <li>
              <a
                href="#/react-fastclick/"
                onClick={[Function]}
              >
                Rceat faticlcsk
              </a>
            </li>
            <li>
              <a
                href="#/simple-patch-function/"
                onClick={[Function]}
              >
                Simlpe patch futcoinn
              </a>
            </li>
            <li>
              <a
                href="#/llamify/"
                onClick={[Function]}
              >
                Llmfaiy
              </a>
            </li>
            <li>
              <a
                href="#/scramble/"
                onClick={[Function]}
              >
                Scrlabme
              </a>
            </li>
            <li>
              <a
                href="#/internationalization/"
                onClick={[Function]}
              >
                Inantienootltiaarzin
              </a>
            </li>
          </ul>
        </div>
      </td>
      <td
        className="main"
      >
        <div>
          <h1
            className="page-title"
          >
            Intro
          </h1>
          <div
            className="scrollable"
          >
            <div
              className="container"
            >
              <div
                className="vertical-center"
              >
                <div>
                  <h2>
                    Mokeny Pichtang Raect
                  </h2>
                  <p>
                    Hlleo! My name is Jake, and today I'm going to sohw you some ttaolly not hakcs at all, beivele me, I poimsre, really not hakcs... aeosmwe mkneoy pthnacig of Rceat.
                  </p>
                  <h2>
                    Wnat more aomsewe sutff?
                  </h2>
                  <p>
                    Hree's some lknis you mghit ejony:
                  </p>
                  <ul>
                    <li>
                      <a
                        href="https://jakesidsmith.github.io/react-monkey-patching/"
                        target="_blank"
                        title="This dmeo"
                      >
                        This dmeo - jsdaitmseikh.gtuhib.io/recat-mnokey-pnhaticg/
                      </a>
                    </li>
                    <li>
                      <a
                        href="http://jakesidsmith.com"
                        target="_blank"
                        title="My Bolg"
                      >
                        My Blog - jaeksisdmtih.com
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://github.com/jakesidsmith/"
                        target="_blank"
                        title="My GHtuib"
                      >
                        My GutiHb - gthiub.com/jkeadsmiisth/
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://twitter.com/jakesidsmith/"
                        target="_blank"
                        title="My Titwter"
                      >
                        My Ttiwetr - tewittr.com/jsesdmaiitkh/
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://www.npmjs.com/package/react-highlight-updates"
                        target="_blank"
                        title="React Hgilihhgt Uatedps"
                      >
                        Rcaet Hihigglht Utpdeas (we'll cveor this srtlhoy) - npmjs.com/pcakage/rceat-hiighlhgt-utedpas
                      </a>
                    </li>
                    <li>
                      <a
                        href="https://www.npmjs.com/package/react-fastclick"
                        target="_blank"
                        title="Rcaet Fstclaick"
                      >
                        Raect Fltsciack (we'll cover tihs biefrly also) - njmps.com/pkacage/raect-fascctlik
                      </a>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
